substrate/frame/support/src/traits/storage.rs

```diff
@@ -127,6 +132,66 @@ impl WhitelistedStorageKeys for Tuple {
 	}
 }
 
+/// The resource footprint of a bunch of blobs. We assume only the number of blobs and their total
+/// size in bytes matter.
+#[derive(Default, Copy, Clone, Eq, PartialEq, RuntimeDebug)]
+pub struct Footprint {
+	/// The number of blobs.
+	pub count: u64,
+	/// The total size of the blobs in bytes.
+	pub size: u64,
+}
+
+impl Footprint {
+	pub fn from_parts(items: usize, len: usize) -> Self {
+		Self { count: items as u64, size: len as u64 }
+	}
+
+	pub fn from_encodable(e: impl Encode) -> Self {
+		Self::from_parts(1, e.encoded_size())
+	}
+}
+
+/// Some sort of cost taken from account temporarily in order to offset the cost to the chain of
+/// holding some data `Footprint` in state.
+///
+/// The cost may be increased, reduced or dropped entirely as the footprint changes.
+///
+/// A single ticket corresponding to some particular datum held in storage. This is an opaque
+/// type, but must itself be stored and generally it should be placed alongside whatever data
+/// the ticket was created for.
+///
+/// While not technically a linear type owing to the need for `FullCodec`, *this should be
+/// treated as one*. Don't type to duplicate it, and remember to drop it when you're done with
+/// it.
+#[must_use]
+pub trait Consideration<AccountId>: Member + FullCodec + TypeInfo + MaxEncodedLen {
```

Comment:
Not sure if the `#[must_use]` attribute put here does anything. I get that the intention is to say that the implementing type must be used in some way, but I don't think the attribute in this position can do so, as the literature didn't document what should happen when placed above a trait item.

Instead, I propose putting it in trait functions that return `Self`, that way we can guarantee that whenever we call the trait function, the result value MUST be used.

Reply:
[Permalink to the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=94526a438756ad2791faf15800b71a79)